src/commands/dev/socket.rs

```diff
@@ -19,66 +30,309 @@ use tokio_tungstenite::{connect_async, tungstenite, MaybeTlsStream, WebSocketStr
+                .map(Ok)
+                .forward(write)
+                .map_err(Into::into);
+
+            let printer = print_ws_messages(read);
+
+            // run the heartbeat and message printer in parallel
+            if tokio::try_join!(heartbeat, keep_alive_to_ws, printer).is_ok() {
+                break Ok(());
+            } else {
+            }
+        }
+    }
+}
 
-        // run the heartbeat and message printer in parallel
-        if tokio::try_join!(heartbeat, keep_alive_to_ws, printer).is_ok() {
-            break Ok(());
+struct WebsocketError {
+    from_chrome: bool,
+    inner: tungstenite::Error,
+}
+
+impl Display for WebsocketError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let direction = if self.from_chrome {
+            "chrome -> edgeworker"
         } else {
+            "edgeworker -> chrome"
+        };
+        write!(f, "{}: {}", direction, self.inner)
+    }
+}
+
+async fn websocket_handle_events(
+    local_stream: WebSocketStream<Upgraded>,
+    remote_stream: &mut WebSocketStream<MaybeTlsStream<TcpStream>>,
+) -> std::result::Result<(), WebsocketError> {
+    info!("opened websocket stream between Chrome and edgeworker");
+    // proxy to the upstream websocket
+    let (local_write, local_read) = local_stream.split();
+    let (remote_write, remote_read) = remote_stream.split();
+    let edgeworker_to_chrome = remote_read
+        .inspect(|message| match message {
+            Ok(m) => debug!("edgeworker -> chrome: {}", m),
+            Err(e) => warn!("{}", e),
+        })
+        .forward(local_write)
+        .map_err(|e| WebsocketError {
+            inner: e,
+            from_chrome: false,
+        });
+    let chrome_to_edgeworker = local_read
+        .filter(|message| {
+            let should_forward = match message {
+                Ok(m) => {
+                    debug!("chrome -> edgeworker: {}", m);
+                    true
+                }
+                Err(e) => {
+                    warn!("{}", e);
+                    // When the user closes the inspect window, chrome will abruptly terminate the
+                    // TCP connection, causing tungstenite to mark the stream as poisoned.
+                    // However, this isn't actually a fatal error: if the user opens a new window,
+                    // Chrome will continue to send packets through the websocket.
+                    // Hide this error from the remote so it doesn't give an error when new packets come it.
+                    !matches!(e, tungstenite::Error::ConnectionClosed)
+                }
+            };
+            std::future::ready(should_forward)
+        })
+        .forward(remote_write)
+        .map_err(|e| WebsocketError {
+            inner: e,
+            from_chrome: true,
+        });
+    try_join!(edgeworker_to_chrome, chrome_to_edgeworker)?;
+    Ok(())
+}
+
+async fn start_websocket(
```

Comment:
Do we do anything special during the WebSocket handshake that requires us to write out the handshake, mostly manually? You can checkout how it's done in the `tails-v2` branch, but there is a higher-level `tokio_tungstenite::connect_async`

Reply:
I didn't know that existed! I'll take a look, thanks :)

(This would be accept_async, but your point still stands.)

Reply:
Hmm, I'm not sure how this would work: `accept_async` requires that I consume the stream, and doesn't let me return an HTTP response - it wants access to the raw TCP socket, which I don't have at this point because this is in a `hyper` request handler. I couldn't get access to the raw socket without avoiding `hyper` altogether, which I'd rather not do since I still have to serve other HTTP requests in the meantime.